scipy/optimize/_minimize.py

```diff
@@ -472,10 +472,23 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
     It should converge to the theoretical solution (1.4 ,1.7).
 
     """
-    x0 = np.asarray(x0)
+    if isinstance(x0, np.matrix) and x0.shape[0] == 1:
+        # flatten matrices for backwards compatibility
+        x0 = np.asarray(x0).squeeze(axis=0)
+    else:
+        x0 = np.asarray(x0)
     if x0.dtype.kind in np.typecodes["AllInteger"]:
         x0 = np.asarray(x0, dtype=float)
 
+    # flatten the input so it is 1d, but wrap the function so that it sees the
+    # original dimensionality. This is primarily for convenience working
+    # with scalars instead of vectors, but generalizes well.
+    orig_fun = fun
+    orig_x0 = x0
+    def fun(x, *args):
+        return orig_fun(x.reshape(orig_x0.shape), *args)
```

Comment:
Although nice, this is not backward-compatible because previously `x0` was cast to 1-d array and passed as such to the function.

Reply:
Once I get this passing, I was considering changing this to:
```python
needs_deprecated_conversion = None if x0.shape == () else False

def fun(x, *args):
    nonlocal needs_deprecated_conversion
    if needs_deprecated_conversion is True:
        return orig_fun(x.ravel(), *args)
    elif needs_deprecated_conversion is False:
        return orig_fun(x.reshape(orig_x0.shape), *args)
    else:
        try:
            f = orig_fun(x.reshape(orig_x0.shape), *args)
            needs_deprecated_conversion = False
            return f
        except Exception:
            try:
                f = orig_fun(x.ravel(), *args)
                emit_warning_to_change_x0_to_be_scalar()
                needs_deprecated_conversion = False
                return f
            except Exception:
                pass  # the function is just broken, reraise the outer error
            raise
```

Reply:
TBH, the `try: ... except: ...` thing looks ugly. I'd prefer to just do `x0 = np.atleast_1d(x0)` at the beginning. I think this particular point is not important for the numdiff changes?

Reply:
Arguably none of the numdiff changes are really needed here - my original plan was to add the reshape wrapping at that level, but it quickly became apparent that wasn't going to work.

One of the motivations of this patch was to remove / deprecate the weird `atleast_1d` behavior that was causing @nschloe  all the pain in the first place

Reply:
Collecting some thoughts:

Most of the optimizers actually do `atleast_1d(x0).flatten()`, so also the dim>1 generalization behavior change looks a bit annoying to pull through.  Again legacy from 2000s... Looks like the change here would then necessarily be hard backward compat break.

Generally, minimize() follows the usual literature picture, where you minimize over a vector of parameters, so `x` is a vector, scalar `x0` gets upcasted to a 1-element one, and the objective function returns something castable to a scalar. For minimizing over just one parameter, I'm not so sure people are consistent with `def f(x): return x**2` and having `x0` be scalar when calling minimize(), because it's a thing that used to not matter, so the proper dimension generalization of `x0` probably won't help with the backward compat issue.

With minimize() objective functions, I think the intent is that any scalar castable things (as in `float(y)`) are ok return values. We probably should just make sure that any internal usage of minimize() is clean, old tests should either get warnings silenced or separate cleanup PR, and new compatibility tests added. For sloppy user code, if Numpy emits warnings, that's probably OK. If Numpy plans to remove the old behavior, because there's likely not small amount of `def f(x): return x**2` style scalar minimization code written, I think it would be better if minimize() is also changed to do the scalar cast as early as possible, and to emulate the old behavior plus a warning as necessary.

Then we have to similarly check if Numpy would break some other of the APIs in a similar way. For most, it probably makes sense to just let the scalar cast warnings to just bubble up.

Since numdiff is not public, for that I think the proper dimension generalization API changes can be done, and probably minimize needs to wrap it such that the scalar cast is done first. Maybe useful to also check if it was also used for computing Hessians...